### DISCUSSION TITLE
Elasticsearch Rollback

### DISCUSSION THREAD

**Discussion by Singularity-Zed** [Security-Onion-Solutions/securityonion#8578] [Unsupported Versions] [answered]
Hey all, recently had to rollback from .140 to .130 due to some stability issues on our distributed air-gapped system. Reverted to a previous snapshot of the manager and all nodes but it seems some changes persisted and are massing with elasticsearch and elastalert. Based upon the logs it seems something >v8.0 managed to survive the rollback, is there a way to manually roll back elasticsearch in order to get this operational again? Need to do so in order to run soup and try .140 again. Any andf all help is appreciated, details below.

Output of so-status
Checking Docker status

    Docker -------------------------------------------------------------------------------------------------------------- [ OK ]

Checking container statuses

    so-aptcacherng ------------------------------------------------------------------------------------------------------ [ OK ]
    so-dockerregistry --------------------------------------------------------------------------------------------------- [ OK ]
    so-elastalert -------------------------------------------------------------------------------------------------- [ MISSING ]
    so-elasticsearch ------------------------------------------------------------------------------------------------- [ ERROR ]
    so-filebeat --------------------------------------------------------------------------------------------------------- [ OK ]
    so-grafana ---------------------------------------------------------------------------------------------------------- [ OK ]
    so-idstools --------------------------------------------------------------------------------------------------------- [ OK ]
    so-influxdb --------------------------------------------------------------------------------------------------------- [ OK ]
    so-kibana ----------------------------------------------------------------------------------------------------------- [ OK ]
    so-kratos ----------------------------------------------------------------------------------------------------------- [ OK ]
    so-logstash --------------------------------------------------------------------------------------------------------- [ OK ]
    so-mysql ------------------------------------------------------------------------------------------------------------ [ OK ]
    so-nginx ------------------------------------------------------------------------------------------------------------ [ OK ]
    so-playbook --------------------------------------------------------------------------------------------------------- [ OK ]
    so-redis ------------------------------------------------------------------------------------------------------------ [ OK ]
    so-sensoroni -------------------------------------------------------------------------------------------------------- [ OK ]
    so-soc -------------------------------------------------------------------------------------------------------------- [ OK ]
    so-soctopus --------------------------------------------------------------------------------------------------------- [ OK ]
    so-telegraf --------------------------------------------------------------------------------------------------------- [ OK ]
    so-wazuh ------------------------------------------------------------------------------------------------------------ [ OK ]`

Output of docker images. Here I notice that elasticsearch and elastalert are both present twice with versions .120 and .130

`REPOSITORY                                             TAG       IMAGE ID       CREATED         SIZE
JFH0:5000/security-onion-solutions/so-soc              2.3.130   cfeec22fb3c0   2 months ago    412MB
JFH0:5000/security-onion-solutions/so-soctopus         2.3.130   6f47882697e4   2 months ago    515MB
JFH0:5000/security-onion-solutions/so-kibana           2.3.130   8d38c490f694   2 months ago    1.5GB
JFH0:5000/security-onion-solutions/so-nginx            2.3.130   6f79b36f294c   2 months ago    129MB
JFH0:5000/security-onion-solutions/so-grafana          2.3.130   2127901d8ae2   2 months ago    585MB
JFH0:5000/security-onion-solutions/so-elasticsearch    2.3.130   e03df236ed30   2 months ago    636MB
JFH0:5000/security-onion-solutions/so-mysql            2.3.130   45b1794e4e60   2 months ago    224MB
JFH0:5000/security-onion-solutions/so-filebeat         2.3.130   b5d689547c62   2 months ago    432MB
JFH0:5000/security-onion-solutions/so-redis            2.3.130   067ddf62f27c   2 months ago    32.3MB
JFH0:5000/security-onion-solutions/so-telegraf         2.3.130   f2c8d01bf42f   2 months ago    153MB
JFH0:5000/security-onion-solutions/so-kratos           2.3.130   f75288c36572   2 months ago    63.8MB
JFH0:5000/security-onion-solutions/so-logstash         2.3.130   eb8825142d01   2 months ago    1.47GB
JFH0:5000/security-onion-solutions/so-wazuh            2.3.130   142441baf1bf   2 months ago    899MB
JFH0:5000/security-onion-solutions/so-playbook         2.3.130   687c3a295c55   2 months ago    605MB
JFH0:5000/security-onion-solutions/so-influxdb         2.3.130   5031cee2d277   2 months ago    183MB
JFH0:5000/security-onion-solutions/so-idstools         2.3.130   53f7c5d4f26f   2 months ago    61.7MB
JFH0:5000/security-onion-solutions/so-elastalert       2.3.130   98e7a3c1b0e8   2 months ago    334MB
JFH0:5000/security-onion-solutions/so-acng             2.3.130   29f8065d70a9   2 months ago    81.9MB
JFH0:5000/security-onion-solutions/so-soc              2.3.120   52f0a81bb300   4 months ago    266MB
JFH0:5000/security-onion-solutions/so-logstash         2.3.120   53fe4fb1824d   4 months ago    1.44GB
JFH0:5000/security-onion-solutions/so-kibana           2.3.120   bcc3356746ea   4 months ago    1.47GB
JFH0:5000/security-onion-solutions/so-elasticsearch    2.3.120   e0d4b1f9b6ac   4 months ago    619MB
JFH0:5000/security-onion-solutions/so-filebeat         2.3.120   39f243b01042   4 months ago    415MB
JFH0:5000/security-onion-solutions/so-nginx            2.3.120   864cbd4f689a   4 months ago    88.4MB
JFH0:5000/security-onion-solutions/so-playbook         2.3.120   28a139282079   4 months ago    605MB
JFH0:5000/security-onion-solutions/so-kratos           2.3.120   d5aa0f4c852c   4 months ago    60MB
JFH0:5000/security-onion-solutions/so-soctopus         2.3.120   14131c8bc73f   4 months ago    500MB
JFH0:5000/security-onion-solutions/so-telegraf         2.3.120   bac84e93466a   4 months ago    153MB
JFH0:5000/security-onion-solutions/so-elastalert       2.3.120   33e709e0f0a8   4 months ago    307MB
JFH0:5000/security-onion-solutions/so-wazuh            2.3.120   cbc8c5a1cebf   4 months ago    899MB
JFH0:5000/security-onion-solutions/so-mysql            2.3.120   6a8c241f3bcb   4 months ago    224MB
JFH0:5000/security-onion-solutions/so-influxdb         2.3.120   b2b695432f70   4 months ago    183MB
JFH0:5000/security-onion-solutions/so-grafana          2.3.120   7d099f9375f4   4 months ago    584MB
JFH0:5000/security-onion-solutions/so-idstools         2.3.120   7b6a5e134b29   4 months ago    61.7MB
JFH0:5000/security-onion-solutions/so-redis            2.3.120   ee668da5ac5d   4 months ago    32.3MB
JFH0:5000/security-onion-solutions/so-acng             2.3.120   82a5bf8a0760   4 months ago    81.9MB
JFH0:5000/security-onion-solutions/so-thehive          2.3.110   0eec2a8ae12d   5 months ago    850MB
JFH0:5000/security-onion-solutions/so-thehive-cortex   2.3.110   0858a5e251ea   5 months ago    1.77GB
JFH0:5000/security-onion-solutions/so-thehive-es       2.3.110   add69acb871c   5 months ago    1.14GB
ghcr.io/security-onion-solutions/registry              latest    b2cb11db9d3d   11 months ago   26.2MB
JFH0:5000/security-onion-solutions/so-thehive-es       2.3.40    83f951bad3d1   17 months ago   1.37GB
JFH0:5000/security-onion-solutions/so-thehive-cortex   2.3.40    02b2d11ab2b1   17 months ago   1.32GB
JFH0:5000/security-onion-solutions/so-thehive          2.3.40    fb8361bd9301   17 months ago   629MB
<none>                                                 <none>    5c4008a25e05   18 months ago   26.2MB
registry                                               latest    2d4f4b5309b1   2 years ago     26.2MB`

Both elasticsearch and elastalert are not present in docker ps

so-elasticsearch outputs

`                  Connection attempt timed out.  Unable to connect to ElasticSearch.
                  Please try:
                    -checking log(s) in /var/log/elasticsearch/
                    -running 'sudo docker ps'
                    -running 'sudo so-elastic-restart'

                  Loading roles...

Summary for local
--------------
Succeeded: 127 (changed=6)
Failed:      0
--------------
Total states run:     127
Total run time:   251.276 s`

Checking the logs in /opt/so/log/elasticsearch/MgrName.log  I see the following.

`[2022-08-23T12:09:37,056][ERROR][org.elasticsearch.bootstrap.Bootstrap] Exception
java.lang.IllegalStateException: data path [/usr/share/elasticsearch/data] is not compatible with Elasticsearch v7.17.4, perhaps it has already been upgraded to a later version
        at org.elasticsearch.env.NodeEnvironment.<init>(NodeEnvironment.java:286) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.node.Node.<init>(Node.java:429) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.node.Node.<init>(Node.java:309) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:234) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:234) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:434) [elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:166) [elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.bootstrap.Elasticsearch.execute(Elasticsearch.java:157) [elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:77) [elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:112) [elasticsearch-cli-7.17.4.jar:7.17.4]
        at org.elasticsearch.cli.Command.main(Command.java:77) [elasticsearch-cli-7.17.4.jar:7.17.4]
        at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:122) [elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:80) [elasticsearch-7.17.4.jar:7.17.4]
Caused by: java.lang.IllegalStateException: [/usr/share/elasticsearch/data/nodes] is a file which contains [written by Elasticsearch v8.3.2 to prevent a downgrade to a version prior to v8.0.0 which would result in data loss]
        ... 13 more
[2022-08-23T12:09:37,062][ERROR][org.elasticsearch.bootstrap.ElasticsearchUncaughtExceptionHandler] uncaught exception in thread [main]
org.elasticsearch.bootstrap.StartupException: java.lang.IllegalStateException: data path [/usr/share/elasticsearch/data] is not compatible with Elasticsearch v7.17.4, perhaps it has already been upgraded to a later version
        at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:170) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.bootstrap.Elasticsearch.execute(Elasticsearch.java:157) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:77) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:112) ~[elasticsearch-cli-7.17.4.jar:7.17.4]
        at org.elasticsearch.cli.Command.main(Command.java:77) ~[elasticsearch-cli-7.17.4.jar:7.17.4]
        at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:122) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:80) ~[elasticsearch-7.17.4.jar:7.17.4]
Caused by: java.lang.IllegalStateException: data path [/usr/share/elasticsearch/data] is not compatible with Elasticsearch v7.17.4, perhaps it has already been upgraded to a later version
        at org.elasticsearch.env.NodeEnvironment.<init>(NodeEnvironment.java:286) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.node.Node.<init>(Node.java:429) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.node.Node.<init>(Node.java:309) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:234) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:234) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:434) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:166) ~[elasticsearch-7.17.4.jar:7.17.4]
        ... 6 more
Caused by: java.lang.IllegalStateException: [/usr/share/elasticsearch/data/nodes] is a file which contains [written by Elasticsearch v8.3.2 to prevent a downgrade to a version prior to v8.0.0 which would result in data loss]
        at org.elasticsearch.env.NodeEnvironment.<init>(NodeEnvironment.java:286) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.node.Node.<init>(Node.java:429) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.node.Node.<init>(Node.java:309) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:234) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:234) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:434) ~[elasticsearch-7.17.4.jar:7.17.4]
        at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:166) ~[elasticsearch-7.17.4.jar:7.17.4]
        ... 6 more`

**Comment by dougburks** [comment]
When updating to 2.3.140 and Elastic 8, did you follow all steps at https://docs.securityonion.net/en/2.3/soup.html#elastic-8?

Under normal circumstances, reverting to a previous snapshot should have reverted the Elasticsearch indices to their Elastic 7 state. However, it appears that your Elasticsearch indices are still in their upgraded Elastic 8 state which might indicate that you have mounted `/nsm` to a remote volume which didn't get reverted like the rest of your system. If that's the case, please be aware that we recommend against remote volumes in our Best Practices. From https://docs.securityonion.net/en/2.3/best-practices.html#installation:
> use local storage and avoid NFS, NAS, iSCSI, etc.

You may be able to recover this installation by removing your Elasticsearch indices and re-initializing some services but it would require some manual work. It may be just as easy to rebuild the system and you may want to do that anyway to more closely align with Best Practices.

**Reply by Singularity-Zed** [reply]
Sadly we did not follow those steps initially, this is the main reason for rolling back to a previous snapshot. Will review best practices and bring up the possibility of a rebuild, I inherited the system from a previous administrator.

If possible it would be preferable to remediate without rebuilding.

**Reply by Singularity-Zed** [reply]
The indices have been removed, what specifically would have to be re-initialized? Ran so-container-restart on both elastalert and elasticsearch but end up with the same results.

logs in /opt/so/log/elasticsearch/MgrName.log still report the following:

> [2022-08-24T12:39:15,945][ERROR][org.elasticsearch.bootstrap.ElasticsearchUncaughtExceptionHandler] uncaught exception in thread [main]
> org.elasticsearch.bootstrap.StartupException: java.lang.IllegalStateException: data path [/usr/share/elasticsearch/data] is not compatible with Elasticsearch v7.17.4, perhaps it has already been upgraded to a later version
>         at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:170) ~[elasticsearch-7.17.4.jar:7.17.4]
>         at org.elasticsearch.bootstrap.Elasticsearch.execute(Elasticsearch.java:157) ~[elasticsearch-7.17.4.jar:7.17.4]
>         at org.elasticsearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:77) ~[elasticsearch-7.17.4.jar:7.17.4]
>         at org.elasticsearch.cli.Command.mainWithoutErrorHandling(Command.java:112) ~[elasticsearch-cli-7.17.4.jar:7.17.4]
>         at org.elasticsearch.cli.Command.main(Command.java:77) ~[elasticsearch-cli-7.17.4.jar:7.17.4]
>         at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:122) ~[elasticsearch-7.17.4.jar:7.17.4]
>         at org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:80) ~[elasticsearch-7.17.4.jar:7.17.4]
> Caused by: java.lang.IllegalStateException: data path [/usr/share/elasticsearch/data] is not compatible with Elasticsearch v7.17.4, perhaps it has already been upgraded to a later version
>         at org.elasticsearch.env.NodeEnvironment.<init>(NodeEnvironment.java:286) ~[elasticsearch-7.17.4.jar:7.17.4]
>         at org.elasticsearch.node.Node.<init>(Node.java:429) ~[elasticsearch-7.17.4.jar:7.17.4]
>         at org.elasticsearch.node.Node.<init>(Node.java:309) ~[elasticsearch-7.17.4.jar:7.17.4]
>         at org.elasticsearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:234) ~[elasticsearch-7.17.4.jar:7.17.4]
>         at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:234) ~[elasticsearch-7.17.4.jar:7.17.4]
>         at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:434) ~[elasticsearch-7.17.4.jar:7.17.4]
>         at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:166) ~[elasticsearch-7.17.4.jar:7.17.4]
>         ... 6 more
> Caused by: java.lang.IllegalStateException: [/usr/share/elasticsearch/data/nodes] is a file which contains [written by Elasticsearch v8.3.2 to prevent a downgrade to a version prior to v8.0.0 which would result in data loss]
>         at org.elasticsearch.env.NodeEnvironment.<init>(NodeEnvironment.java:286) ~[elasticsearch-7.17.4.jar:7.17.4]
>         at org.elasticsearch.node.Node.<init>(Node.java:429) ~[elasticsearch-7.17.4.jar:7.17.4]
>         at org.elasticsearch.node.Node.<init>(Node.java:309) ~[elasticsearch-7.17.4.jar:7.17.4]
>         at org.elasticsearch.bootstrap.Bootstrap$5.<init>(Bootstrap.java:234) ~[elasticsearch-7.17.4.jar:7.17.4]
>         at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:234) ~[elasticsearch-7.17.4.jar:7.17.4]
>         at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:434) ~[elasticsearch-7.17.4.jar:7.17.4]
>         at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:166) ~[elasticsearch-7.17.4.jar:7.17.4]

The directory listed in the logs, /usr/share/elasticsearch does not exist. Is there something I would have to remove other than the indices?

**Comment by Singularity-Zed** [comment, marked as answer]
I was able to figure this out. Combining the logs and your comment about the /nsm I found the nodes file referenced in the log in /nsm/elasticsearch/nodes. Once I removed that I was able to restart the container and get it operational in .130 again. Not sure why the log was referencing the wrong location but oh well

Thank you for sending me in the right direction. Log in question below for reference

>Caused by: java.lang.IllegalStateException: [/usr/share/elasticsearch/data/nodes] is a file which contains [written by Elasticsearch v8.3.2 to prevent a downgrade to a version prior to v8.0.0 which would result in data loss]

**Reply by dougburks** [reply]
> Not sure why the log was referencing the wrong location 
 [/usr/share/elasticsearch/data/nodes]

Elasticsearch runs in a container. When the container starts, it binds `/nsm/elasticsearch` on the host to `/usr/share/elasticsearch/data` inside the container.

Since this system was inherited and didn't follow Best Practices, we highly recommend that if you have any other problems to please consider rebuilding following the Best Practices:
[https://docs.securityonion.net/en/2.3/best-practices.html#installation](https://www.google.com/url?q=https://docs.securityonion.net/en/2.3/best-practices.html%23installation&sa=D&source=calendar&ust=1661805066096259&usg=AOvVaw3B1fVy-DDEzYQXflWiHZfk)

**Reply by Singularity-Zed** [reply]
Understood, thank you for clarifying. Definitely considering a rebuild
